src/Datadog.Trace/Agent/NamedPipes/TraceRequest.cs

```diff
@@ -0,0 +1,78 @@
+using System.Collections.Generic;
+using Datadog.Trace.Vendors.MessagePack;
+
+namespace Datadog.Trace.Agent.NamedPipes
+{
+    internal class TraceRequest
+    {
+        public Dictionary<string, string> Headers { get; set; } = new Dictionary<string, string>();
+
+        public Span[][] Traces { get; set; }
+
+        internal int WriteHeaders(ref byte[] bytes, int offset)
+        {
+            int originalOffset = offset;
+
+            offset += MessagePackBinary.WriteString(ref bytes, offset, "headers");
+
+            int count = 0;
+
+            if (Headers != null)
+            {
+                lock (Headers)
```

Comment:
Is there a situation where the headers could be modified while the trace is serialized? That sounds scary

Reply:
I don't know, but it seemed safe to do so. This is mostly a carryover from other methods.